### ISSUE TITLE
Link support

### ISSUE THREAD

**Issue by yaxu** [tidalcycles/Tidal#660]
It'd be good to have link support natively in Tidal/Haskell. An alternative is relying on SuperCollider's new link support, but [as is noted here](https://github.com/musikinformatik/SuperDirt/issues/185), that's not straightforward.

This could help: https://github.com/wavewave/fficxx

**Comment by yaxu** [comment]
Once again, I've decided that interfacing between Haskell and C++ is completely beyond me. If someone with better understanding of this can help, that'd be great.

**Comment by telephon** [comment]
There is link support in supercollider, but I am not sure if sending messages to tidalcycles from there would help.

**Comment by yaxu** [comment]
It's a bit tricky I think @telephon https://github.com/musikinformatik/SuperDirt/issues/185#issuecomment-600688196

IIRC the current support via a python bridge works by continually checking the system clock time and sending messages relative to that, but I think it's not a perfect approach.

**Comment by telephon** [comment]
yes, indeed.

**Comment by grammoboy2** [comment]
Not sure if this is useful and it has not a proper release yet but...

https://github.com/falkTX/Hylia/blob/master/hylia.h

"Basically the idea was to "dumb down" the link stuff to a super simple C API, which is easier to understand than the official C++ templated classes. Plus also making it easier to build, just needs a "make"

It's used in the Carla project https://github.com/falkTX/Carla

**Comment by grammoboy2** [comment]
> Once again, I've decided that interfacing between Haskell and C++ is completely beyond me. If someone with better understanding of this can help, that'd be great.

Maybe there are options which are easier to implement in both SuperCollider and Tidal?

I saw:

https://www.semanticscholar.org/paper/Simple-Synchronisation-for-Open-Sound-Control-Madgwick-Mitchell/fa139d63858ed49698153486191a29f181ce2e43

AES67, Miracast/Wi-Fi Direct/TDLS, ... ? (not that I've any technical knowledge about these... ;) )

**Comment by yaxu** [comment]
Thanks @grammoboy2 I've made a start on making haskell bindings for hylia. Help welcome!
https://github.com/yaxu/hylia-haskell

**Comment by Zalastax** [comment]
Hi @yaxu and @fbous, as discussed in https://github.com/yaxu/hylia-haskell/issues/1, we should continue the discussion on how to proceed here. I just managed to get some very simple Ableton link integration working over at https://github.com/Zalastax/Tidal/tree/tidal-link. It is very inspired by Hylia and https://github.com/yaxu/hylia-haskell but done from scratch. Next step will be to look at https://github.com/supercollider/supercollider/blob/develop/lang/LangPrimSource/SC_LinkClock.cpp for inspiration as suggested in https://github.com/falkTX/Hylia/issues/2.

**Comment by yaxu** [comment]
@Zalastax Nice one, I tried building under linux out of curiosity, it didn't work yet but I think that won't be a surprise.

There are some things like being able to pause the clock that have been broken for some time, it'll be great to get that working again in this process!

**Comment by Zalastax** [comment]
I've been on a wild goose hunt this last week.

I had a failure to use `cabal v1-install` for my new tidal library. The error was `installed package info from too old version of Cabal (key field does not match id field)`. This caused me to reinstall with many different versions of Haskell and Cabal. Turns out I introduced this error by naming my library...
```
library lib
  ghc-options: -Wall
```

I named my library because I found that `cabal v1-repl` crashes when I try to do `this->link.enable(true);`. So I went and created an executable and somewhere in that process I introduced the name to the library when that was not actually needed. Many hours later I can both produce an executable and install the library. The executable runs fine - `this->link.enable(true);` works and I can see that the beats start ticking. But `cabal v1-repl` still crashes and if I load my installed module in ghci it crashes when I try to enable link as well. I don't know how to debug this - from what I can read it is quite cumbersome to load ghci in gdb. My best guess right now is that there is some DLL that I need to load in GHCI.

The difficulty of debugging this is a concern to me. I don't know if we could keep the C++ parts in a separate executable instead or if the communication overhead would make it infeasible.

**Comment by Zalastax** [comment]
Progress!

I switched to building the C++ parts into a shared library which I build separately. The wrapper for `this->link.enable(true);` now works in both GHCi and in my built executable. I had to build the shared library with flags `-static-libgcc -static-libstdc++` or I would get a runtime error.

My suggestion is to continue with shared library over direct FFI. I'll prepare a new small patch as soon as I have tidied it up a bit!

**Comment by yaxu** [comment]
Good stuff! I don't know if this is relevant or not, but with some help I found building shared libraries helped with making haskell binaries with working haskell interpreters https://github.com/haskell-hint/hint/issues/64

**Comment by Zalastax** [comment]
Thanks for the comment - it gave some comfort that I was on the right track 🙏 

https://github.com/Zalastax/Tidal/tree/tidal-link/tidal-link now contains code that builds a shared library using `Setup.hs` and places it in the cabal bin directory. The beat time is correct and shows up + syncs with the LinkHut application from Ableton 😄 I hope this is the last time fighting Cabal for a while 😅

**Comment by yaxu** [comment]
As a proof of concept, tidalvortex has a simple link-based clock: https://codeberg.org/TidalCycles/vortex/src/branch/main/py-vortex/py_vortex/stream.py#L74
What bothers me is the use of osc bundle timestamps relative to the ntp epoch rather than link clock.. It requires continually checking link time against system clock time and still relying upon them being synched between client and server, which is a problem if the processes are running on different systems. 
But if we support sending link's global host time in an osc message, the server can ignore any ntp timestamp and use that instead.
Hopefully that's easy enough to support in SuperDirt @telephon ? If e.g. link is active in sclang, and a `linktime` parameter is in a message, use that to schedule instead of any bundle timestamp?

**Comment by yaxu** [comment]
Looks like link now has a C wrapper of its own, maybe that simplifies things a little @Zalastax https://github.com/Ableton/link/issues/71#issuecomment-1005686048 ?

**Comment by Zalastax** [comment]
Hi @yaxu, I'll check out the vortex example! The C wrapper from link seems very similar to my own C++ wrapper. I'll not switch to it now but it can be considered as part of clean-up. So far I have focused on refactoring the "clocked" process to be responsible for all Tempo state editing instead of having Stream do that as well. But that is almost done and the next step is to figure out how to actually sync with Link. I hope you will agree with the refactoring 😺 Based on what you write about Vortex, I take it that I need to look into how SuperDirt understands timestamps as well to deliver something really good?

**Comment by yaxu** [comment]
Cool looking forward to having a look!

Yes SuperDirt undertands timestamps according to the OSC standard, which involves putting each message inside a bundle, which has a ntp-style timestamp. I think in practice this isn't a big problem as long as tidal and superdirt are running on the same computer, so we could look at that later.

**Comment by Zalastax** [comment]
It's uploaded now at https://github.com/Zalastax/Tidal/tree/tidal-link!

Next step is to do something similar as you have done in Vortex by start using Link timeAtBeat instead of Tempo cyclesToTime and to go from Link time to OSC time. I expect the trickiest obstacle to be handling resetCycles.

**Comment by yaxu** [comment]
Great, could you create a PR please to make it a bit easier to browse the changes?

**Comment by Zalastax** [comment]
Done!

**Comment by Zalastax** [comment]
Done @yaxu !

**Comment by Zalastax** [comment]
I noticed that my wrapper would need to get quite a bit larger so I went with your suggestion to use Link's C-wrapper. That is in place now and I've started using hsc2hs. Current commit breaks the integration I had with Tidal but I will fix it soon.

The reason this took a while is that at the same time as above, I tried to get rid of the DLL. I did not succeed with that so I have created an issue in GHC https://gitlab.haskell.org/ghc/ghc/-/issues/20918.

**Comment by Zalastax** [comment]
That sounds really cool @yaxu! Unfortunately, it's not ready yet. I think I know exactly what to do and it seems quite straightforward. I just need to find a few hours to make the changes to processCps and toOSC. If I'm lucky, this weekend will be enough - I need to find the hours and debugging needs to go well. I'll ping you when it's working!

Do you need resetCycles to work for your test? It can't be done via Tidal alone so I'll need to add back that part of the tempo sync server before we merge.

**Comment by yaxu** [comment]
@Zalastax Ok no stress but it'd be great to try it. I don't use resetCycles (as currently implemented, it's not accurate)

**Comment by telephon** [comment]
> Hopefully that's easy enough to support in SuperDirt @telephon ? If e.g. link is active in sclang, and a `linktime` parameter is in a message, use that to schedule instead of any bundle timestamp?

I haven't used `LinkClock` in sclang so far, so I am not sure. But I would expect that if you send the time in beats, then sclang can call `schedAbs(beats, item)` and it will end up on the correct beat in the link frame of reference.

As such, this is easy to implement, it can even be plugged in at first without any code change, but people will need to test it well.

**Comment by grammoboy2** [comment]
@jamshark70 knows a lot about Link in SuperCollider.

**Comment by yaxu** [comment]
Thanks for the updates, sadly I haven't got time to get it working under Linux on the raspberry pis for the installation I'm doing on Tuesday but can have a go after that.

**Comment by grammoboy2** [comment]
Thanks for the work so far. Any progress lately? It's ready when it's ready I guess, but it would be quite nice to have.  :)

**Comment by yaxu** [comment]
@grammoboy2 yes there is a PR ready! Please help test it then we can merge :)
https://github.com/tidalcycles/Tidal/pull/898

**Comment by f00b455** [comment]
> The time that I've time to test comes closer. ;) How to build with the particular patch? Any specific things to test?

I would like to test as well, but i would need detailed build instructions (macOs)

**Comment by grammoboy2** [comment]
@Zalastax @yaxu, let us know if, when and how we can help testing please. Cheers.

**Comment by ahihi** [comment]
i am very interested in this functionality as well! i did some quick testing on mac, maybe this helps you @grammoboy2 @f00b455:

```
git clone https://github.com/Zalastax/Tidal.git
cd Tidal
git checkout tidal-link
cabal v1-install --force-reinstalls . ./tidal-link/
```

(this replaced my existing tidal installation, sort of? the old package still shows up `ghc-pkg list`, but in red)

now start tidal, and in the ghci session run:

```
Sound.Tidal.Link.hello
```

then i launched my daw (Bitwig), turned on Link sync and it worked! tempo changes propagate both ways. the sync is slightly off however, with Tidal audibly a little late compared to Bitwig. is there some way to tune that?

**Comment by grammoboy2** [comment]
Thanks

This might help with the latency:

https://tidalcycles.org/docs/configuration/multiuser-tidal/#link-protocol-synchronization
https://scsynth.org/t/ableton-link-latency/2484
https://www.youtube.com/watch?v=rdc-uPfkSCQ

**Comment by f00b455** [comment]
An adjustable latency compensation like 
`Sound.Tidal.Link.hello -0.14`
similar to  `sock <- carabiner tidal 4 (-0.14)`  
would be very nice :-)

**Comment by grammoboy2** [comment]
Are you with us? :) @Zalastax

**Comment by grammoboy2** [comment]
As far as I understand, SuperCollider (SC) does adjust that latency automatically. Tidal uses SC. Tidal sends OSC messages to the SCserver, just like SClang? 
What if you tune that Link latency in Bitwig? @f00b455

**Comment by f00b455** [comment]
As far as i understood the code from @Zalastax , AbletonLink implementation is used directly from Tidal. I experienced that it is better, to do the latency compensation there where it emerges, because every link client can have, for example different buffer sizes, resulting in different latencies.

I implemented AbletonLink in AudioStellar with a slider in the prefs menu for latency compensation.
https://gitlab.com/ayrsd/audiostellar/-/blob/units/src/Servers/AbletonLinkServer.cpp


![image](https://user-images.githubusercontent.com/920683/169529513-2827ca04-e388-4a63-afd0-7b624f8112de.png)

**Comment by grammoboy2** [comment]
Ok, yeah my technical knowledge is too limited here. Others need to chime in.

**Comment by telephon** [comment]
> As far as I understand, SuperCollider (SC) does adjust that latency automatically. Tidal uses SC. Tidal sends OSC messages to the SCserver, just like SClang?

Yes, no need to do that separately. You just send the intended time, sc will do the rest. But all the clock syncing will be better done in tidalcycles.

**Comment by jamshark70** [comment]
> As far as I understand, SuperCollider (SC) does adjust that latency automatically.

A bit more detail about that.

Ableton defines Link time values as the moment when the corresponding audio should go out to the hardware ("hit the speakers").

SC server defines OSC timestamps the same -- the moment of hitting the speakers. (That is, it's scsynth's responsibility to compensate for the hardware buffer size.)

Link runs in sclang. Everything in sclang happens at a logical time on one or another clock: Wake up at beat 100, logical time = 100. The OSC timestamp comes from converting this logical time (which may be in beats) into physical seconds, then adding a latency value to absorb network jitter (by default, 0.2 seconds). With a non-Link TempoClock, this means that the audio you hear is later than clock time -- but it doesn't matter because everything is later by the same amount, and (since TempoClock is only in sclang) there's no reference to any other timebase.

With LinkClock, there *is* a reference to another timebase: Link's.

SC can't get rid of messaging latency. So we have to run the language-side clock early: If beat 100's sound should hit the speakers at Link time 100 (i.e. at quantum = 4, phase 0), then sclang must wake up to the scheduled item at beats2secs(100) - server_latency -- the outgoing OSC bundle adds server_latency back in.

To do this, we set a property in the clock: `TempoClock.default = LinkClock.new.latency_(s.latency);`.

I'm not familiar with Tidal's internals, but I suppose it wouldn't be outrageous to handle it in a similar way.

**Comment by f00b455** [comment]
> A bit more detail about that.
> 
> Ableton defines Link time values as the moment when the corresponding audio should go out to the hardware ("hit the speakers").
> 
I am not sure. Ableton Link just shares a timeline over IP and does not know anything about the implementing software latency issues.

> 
> Link runs in sclang. Everything in sclang happens at a logical time on one or another clock: Wake up at beat 100, logical time = 100. The OSC timestamp comes from converting this logical time (which may be in beats) into physical seconds, then adding a latency value to absorb network jitter (by default, 0.2 seconds). With a non-Link TempoClock, this means that the audio you hear is later than clock time -- but it doesn't matter because everything is later by the same amount, and (since TempoClock is only in sclang) there's no reference to any other timebase.

As far as i understood, is the goal to implement Ableton Link natively over c++ bindings in Tidal. So you have to deal with latencies:
Tidal -> OSC -> SuperCollider -> SuperDirt -> Buffer -> Interface -> Speakers

**Comment by f00b455** [comment]
https://media.ccc.de/v/lac2018-42-ableton_link_a_technology_to_synchronize_music_software#t=1300

**Comment by Zalastax** [comment]
Sorry for being missing in action. As we now have a solution for https://gitlab.haskell.org/ghc/ghc/-/issues/20918, I will engage actively with this again. As it is a patch on GHC master, many of our dependencies are broken. I will try to use https://gitlab.haskell.org/ghc/head.hackage/-/tree/master/patches to circumvent that. My goal right now will be to get REPLs working on Windows, e.g. in Atom / VS Code.

Regarding where to invoke Link to convert from beats to clock, moving it from Tidal to SuperCollider would probably only have minor advantages over adjusting the latency in SuperCollider in the way proposed by @jamshark70 `TempoClock.default = LinkClock.new.latency_(s.latency);`. `send` in `Stream.hs` could easily be adopted to send beats instead of time in case we want to try it out anyway.

@ahihi, thanks for spotting that! I will have a look and try to fix it.

Feel free to reach out to me on Discord if you need help testing the Pull Request. I have the same handle: @Zalastax.

**Comment by Zalastax** [comment]
I managed to get most dependencies to compile but [Hint](https://github.com/haskell-hint/hint/) is proving very challenging. It is closely coupled to GHC. I am going offline next week but I'll see if I can create a patch for Hint when I am back. If it proves too challenging, I will open an issue instead.

**Comment by Zalastax** [comment]
@ahihi I have pushed a correction of the issue with loopAt. Thanks for noticing!

https://github.com/haskell-hint/hint/issues/150 has been opened now 😺

**Comment by Zalastax** [comment]
I believe the PR is ready to be merged now. I opted for removing targets that use hint when compiling on Windows. Please let me know your verdicts!

**Comment by Zalastax** [comment]
Link support is now merged 🥳 Now it's time to update the documentation.

**Comment by f00b455** [comment]
@Zalastax - Nice one ant thanks a lot! i got it working, but i do have audible latency. Tidal is late. Is there any way to adjust this?

**Comment by Zalastax** [comment]
Hi @f00b455, are you seeing any `late` messages in Supercollider? Supercollider *should* schedule the notes correctly according to the calculated OSC time, but perhaps the latency parameter needs to be adjusted https://composerprogrammer.com/teaching/supercollider/sctutorial/8.1%20Precise%20Timing%20in%20SC.html.  Also see https://scsynth.org/t/ableton-link-latency/2484/2. We will need to create a knowledge base around this so we can write good documentation. We can also again consider adjusting how we communicate time to SuperDirt if other changes don't work out. Feel free to contact me on Discord!

**Comment by jamshark70** [comment]
> Nice one and thanks a lot! i got it working, but i do have audible latency. Tidal is late. Is there any way to adjust this?

Scsynth audio depends, for consistent timing, on timestamps calculated as the current logical time plus a fixed number of seconds ("server latency").

This means *all* audio is late by that amount. If SC's timebase is the only active timebase, then this isn't a problem -- you don't hear the language-side clock, and everything is late together, thus sounds together.

With Link, SC is not the only timebase, and the consistent lateness becomes apparent.

Ableton Link establishes a relationship between a number of beats and a physical time point, let's say in seconds. Sclang's LinkClock just subtracts the latency from the number of seconds: if server latency is 0.2, and Link says that beat 100 is at time 162.547, then SC will wake up beat 100 at 162.347 seconds. Then server messages are scheduled for 0.2 seconds later, and it sounds on time.

If scsynth is providing the audio, then Tidal Link support will need to emulate this approach. I really don't see any other way. (I see that I talked about this before, but it isn't clear from the issue history whether it was actually done.)

The composerprogrammer link suggests reducing the latency value. You cannot take it down to zero (you will definitely get late messages that way!). This will not give you good sync with other Link peers -- SC/Tidal audio will be *slightly* late, and inconsistently late.

**Comment by Zalastax** [comment]
@jamshark70, the approach taken so far is to send OSC messag bundles that use exactly the time calculated by Link. The OSC messages are calculated ahead of time so that SC shouldn't produce any 'late' messages - how much ahead is configurable. I might have misunderstood what the meaning of the time put in the bundle means. I can add code that subtracts time from the bundle if that's the right thing to do.

**Comment by jamshark70** [comment]
> I see you suggesting to use TempoClock.default = LinkClock.new.latency_(s.latency);

No, that's an explanation of how SC does it. If Tidal is in control of sequencing, then Tidal should connect to Link (and there'd be no need for SC to connect).

> The OSC messages are calculated ahead of time so that SC shouldn't produce any 'late' messages - how much ahead is configurable.

Yes, this is mimicking SC's TempoClock (non-Link) usage.

> I might have misunderstood what the meaning of the time put in the bundle means. I can add code that subtracts time from the bundle if that's the right thing to do. Should the time put in the bundle be subtracted by 0.2s by default?

No. The timestamp needs to be the moment when you want the sound to come out of the speakers.

Tidal needs to run its sequencing logic *earlier* than Link peers, and then the latency added into the timestamp compensates so that it sounds on time. In SC's codebase, BeatsToSecs() subtracts latency, SecsToBeats() adds it:

```
double LinkClock::BeatsToSecs(double beats) const {
    auto sessionState = mLink.captureAppSessionState();
    double secs = linkToHrTime(sessionState.timeAtBeat(beats, mQuantum)) - mLatency;
    return secs;
}

double LinkClock::SecsToBeats(double secs) const {
    auto sessionState = mLink.captureAppSessionState();
    double beats = sessionState.beatAtTime(hrToLinkTime(secs + mLatency), mQuantum);
    return beats;
}
```

I guess this is likely to be in Sound.Tidal.Tempo but I don't know Tidal's codebase.

**Comment by Zalastax** [comment]
I held the belief that the OSC bundle time was signifying at what time the sound should hit the speaker. I therefore didn't see a need to account for latency when calculating the OSC bundle time. After reading [the standard](https://ccrma.stanford.edu/groups/osc/spec-1_0.html) it instead seems to be the time Supercollider will invoke SuperDirt:

> When a received OSC Packet contains only a single OSC Message, the OSC Server should invoke the correponding OSC Methods immediately, i.e., as soon as possible after receipt of the packet. Otherwise a received OSC Packet contains an OSC Bundle, in which case the OSC Bundle's OSC Time Tag determines when the OSC Bundle's OSC Messages'
corresponding OSC Methods should be invoked. If the time represented by the OSC Time Tag is before or equal to the current time, the OSC Server should invoke the methods immediately (unless the user has configured the OSC Server to discard messages that arrive too late). Otherwise the OSC Time Tag represents a time in the future, and the OSC server
must store the OSC Bundle until the specified time and then invoke the appropriate OSC Methods.

With this knowledge, it becomes clear that it's correct to do as @jamshark70 says and account for latency when producing the OSC bundle. The latency to account for should be the latency of SuperDirt or the MIDI device, etc. I will add logic for this in the coming days. It would be neat if it was possible to use different latency for a MIDI device and SuperDirt running in the same Supercollider, but that seems difficult so let's start with one latency value per target.

**Comment by yaxu** [comment]
Yes I don't like that part of the OSC spec, some libraries implement it by default and it causes a lot of confusion. Best to ignore it and follow the supercollider way indeed.

Supercollider does have a separate midi latency adjustment which is handy for getting everything in phase.

**Comment by yaxu** [comment]
I disagree with the bit that says that the OSC receiver should pause when it receives a bundle with a timestamp. Superdirt rather processes it right away and leaves the pausing for synthesis scheduling.

**Comment by f00b455** [comment]
As a user, i would like to be able to move the scheduling of tidal, back and foreward on the AbletonLink timelime. This would enable me, to adjust hardware latencies in a heterogeneous setup, with virtual audio/MIDI, hardware synth, Ableton Live, audio hardware, etc. Is this already possible, or any workaround?

**Comment by Zalastax** [comment]
Hi @f00b455, I plan to update the code at https://github.com/tidalcycles/Tidal/blob/7b1192b95da180378e4b2ce077e92c41fc310e9c/src/Sound/Tidal/Stream.hs#L548-L566
I think the code will look like this afterwards:
```Haskell

-- send has three modes:
-- Send events early using timestamp in the OSC bundle - used by Superdirt
-- Send events early by adding timestamp to the OSC message - used by Dirt
-- Send events live by delaying the thread
send :: Maybe O.UDP -> Cx -> Double -> Double -> (Double, Bool, O.Message) -> IO ()
send listen cx latency extraLatency (time, isBusMsg, m)
  | oSchedule target == Pre BundleStamp = sendBndl isBusMsg listen cx $ O.Bundle (timeWithLatency) [m]
  | oSchedule target == Pre MessageStamp = sendO isBusMsg listen cx $ addtime m
  | otherwise = do _ <- forkOS $ do now <- O.time
                                    threadDelay $ floor $ (timeWithLatency - now) * 1000000
                                    sendO isBusMsg listen cx m
                   return ()
    where addtime (O.Message mpath params) = O.Message mpath ((O.int32 sec):((O.int32 usec):params))
          ut = O.ntpr_to_ut timeWithLatency
          sec :: Int
          sec = floor ut
          usec :: Int
          usec = floor $ 1000000 * (ut - (fromIntegral sec))
          target = cxTarget cx
          timeWithLatency = time - latency + extraLatency
```

Feel free to test this change and to open a Pull Request if it works!
Your use case should after the change be possible to achieve by changing oLatency in https://github.com/tidalcycles/Tidal/blob/9db202d775458dcfec4b629e6cb9114829fa3e90/BootTidal.hs#L9

**Comment by f00b455** [comment]
Hi @Zalastax ,

i was not able to adjust the latency. I did following steps:

- cloned from your repository (main)
- changed in tidal.cabal the version to 1.8.1
-`cabal v1-install --force-reinstalls`
- started atom 
- <img width="725" alt="image" src="https://user-images.githubusercontent.com/920683/179233094-1263254f-6bb2-47ad-9ce6-0e81b749c5a0.png">
- changed oLatency in BootTidal.hs

**Comment by Zalastax** [comment]
Hi @f00b455, please try main of the Tidal repository. My main branch is outdated. Please contact me on Discord if you still are stuck after that.

**Comment by f00b455** [comment]
Hi @Zalastax , nice one! This WORKS GREAT! Thank you ALL!

**Comment by yaxu** [comment]
Yes this is lovely !

Playing with dynamic cps with two tidal instances running on the same computer, I noticed that e.g. `cps "<0.75 0.5 0.125>"` worked well, but `d2 $ slow 8 $ segment 16 $ cps (range 0.5 3 saw)` running on one instance caused glitches in the second instance like missed and doubled events.

**Comment by Zalastax** [comment]
Good experiment @yaxu!

Did you see any `skip: ` messages? I have tried to design the logic so that skipped or repeated events can only happen if there is a large time drift. `skip: ` is printed in Time.hs when drift is detected. If you don't see it, there might be a bug somewhere.

I can try reproducing during next week.

**Comment by yaxu** [comment]
Am I right that with this merged, windows users must have a particular, recent version of ghc to install tidal?

**Comment by yaxu** [comment]
> Did you see any `skip: ` messages? I have tried to design the logic so that skipped or repeated events can only happen if there is a large time drift. `skip: ` is printed in Time.hs when drift is detected. If you don't see it, there might be a bug somewhere.

No I didn't get skip messages, as far as I remember. It would be great to see this improved, but TBH I would have been surprised if the timing had been _perfect_ with this stress test, at least the instance changing the cps had smooth changes.

**Comment by Zalastax** [comment]
Yes @yaxu, [GHC 9.4.1](https://www.haskell.org/ghc/blog/20220807-ghc-9.4.1-released.html) is required on Windows.

**Comment by yaxu** [comment]
> Yes @yaxu, [GHC 9.4.1](https://www.haskell.org/ghc/blog/20220807-ghc-9.4.1-released.html) is required on Windows.

Hm I'm trying to install with 9.4.1 (under linux), and am getting all kinds of dependency problems, @aravind-mohandas seems to have the same issue under windows.. and the windows github actions seem to be using ghc 9.0.1

**Comment by Zalastax** [comment]
Indeed. We need to wait a bit until dependencies are updated. I used `cabal v2-install --lib . ./tidal-link/ --allow-newer=ghc,template-haskell,text,base` to install on my machine.

It is possible to build on Windows with GHC 9.0.1 but it will crash when you run Tidal. GHC 9.4.1 is the first release that includes the necessary bug fixes. I hope this clarifies!

**Comment by yaxu** [comment]
@Zalastax Thanks, I updated the upper bounds in the .cabal files so maybe the allow-newer isn't needed any more. (`cabal outdated` tells you what to do)

**Comment by aravind-mohandas** [comment]
Hi, I've been trying to get tidal-1.9 to build on Windows, using ghc - 9.4.1. I keep running into this error during the build, 

![image](https://user-images.githubusercontent.com/59365578/185115691-c38c71a4-8f07-4716-ae35-0b7636b8e023.png)

and once it goes to ghc 9.2.1, then it has a problem with the requirements regarding base.

I have been having the same issue in getting my [Haskell Summer of Code] (https://github.com/tidalcycles/konnakol-gsoc) to build, but there it runs into the same error even when building Tidal-1.8. 

I would appreciate any help 😄

**Comment by yaxu** [comment]
Hi @aravind-mohandas , this is failing due to the 'hint' library not yet working with ghc 9.4.1. However, the tidal library doesn't have hint as a dependency, so this shouldn't be a problem. There _is_ an tidal executable that does use hint, but that shouldn't be built if you specify `cabal install --lib`. So probably this is a bug in cabal.

Anyway, one workaround for this is to edit tidal.cabal to remove or comment out the whole 'executable tidal' section.

**Comment by Zalastax** [comment]
Hi @aravind-mohandas, can you contact me on Discord? Same username: Zalastax.

The first thing I would suggest is to check if the libc++.dll that is distributed with GHC is on your PATH.

**Comment by Zalastax** [comment]
https://github.com/tidalcycles/Tidal/pull/944 is now merged. @aravind-mohandas, can you please see if your issue is resolved? I no longer need to have `C:\tools\ghc\mingw\bin;` on my PATH.

**Comment by aravind-mohandas** [comment]
Yes, it has been resolved. Thank you!

**Comment by mindofmatthew** [comment]
Just sorting through old issues—is this resolved now with the 1.9 releases?

**Comment by Zalastax** [comment]
Yes :)
